odl/solvers/functional/default_functionals.py

```diff
@@ -2470,6 +2470,149 @@ def __repr__(self):
+                                          '`functional` {} does not implement '
+                                          '`functional.gradient`'
+                                          ''.format(functional))
+        else:
+            # Check that given subgradient is an operator that maps from the
+            # domain of the functional to the domain of the functional
+            if not isinstance(subgradient_op, Operator):
+                raise TypeError('`subgradient_op` {} is not an instance of '
+                                '``Operator``'.format(subgradient_op))
+            if not self.__functional.domain == subgradient_op.domain:
+                raise ValueError('`functional.domain` {} is not the same as '
+                                 '`subgradient_op.domain` {}'
+                                 ''.format(self.__functional.domain,
+                                           subgradient_op.domain))
+            if not self.__functional.domain == subgradient_op.range:
+                raise ValueError('`functional.domain` {} is not the same as '
+                                 '`subgradient_op.range` {}'
+                                 ''.format(self.__functional.domain,
+                                           subgradient_op.range))
+            self.__subgradient_op = subgradient_op
+
+        self.__subgrad_eval = self.__subgradient_op(self.__point)
+        self.__bregman_dist = (FunctionalQuadraticPerturb(
+            self.__functional, linear_term=-self.__subgrad_eval) -
+            self.__functional(point) + self.__subgrad_eval.inner(point))
+
+        super(BregmanDistance, self).__init__(
+            space=functional.domain, linear=False,
+            grad_lipschitz=self.__functional.grad_lipschitz)
+
+    @property
+    def functional(self):
+        """The underlying functional for the Bregman distance."""
+        return self.__functional
+
+    @property
+    def point(self):
+        """The underlying point for the Bregman distance."""
+        return self.__point
+
+    @property
+    def subgradient_op(self):
+        """The underlying subgradient operator for the Bregman distance."""
+        return self.__subgradient_op
+
+    def _call(self, x):
+        """Return ``self(x)``."""
+        return self.__bregman_dist(x)
+
+    @property
+    def convex_conj(self):
+        """The convex conjugate"""
+        return self.__bregman_dist.convex_conj
+
+    @property
+    def proximal(self):
+        """Return the ``proximal factory`` of the functional."""
+        return self.__bregman_dist.proximal
+
+    @property
+    def gradient(self):
+        """Gradient operator of the functional."""
+        return self.subgradient_op - ConstantOperator(self.__subgrad_eval)
+
+    def __repr__(self):
+        '''Return ``repr(self)``.'''
+        return '{}({!r}, {!r}, {!r}, {!r})'.format(self.__class__.__name__,
+                                                   self.domain,
```

Comment:
This argument is not in __init__

Reply:
It's more maintainable to use the helper functions for this:
```python
posargs = [functional, point]
subgrad_op_default = getattr(self.functional, 'gradient', None)
optargs = [('subgradient_op', self.subgradient_op, subgrad_op_default)]
inner_str = signature_string(posargs, optargs, sep=',\n')
return '{}(\n{}\n)'.format(self.__class__.__name__, indent(inner_str))
```
The utilities `signature_string` and `indent` are in `odl.util`.
(BTW, I'm working on making this MUCH better).